### PR TITLE
Add Attribute.alias

### PR DESCRIPTION
**Pull request by asford**
# Summary

[PEP681](https://peps.python.org/pep-0681/) introduces an optional `alias` parameter to field descriptors, which specifies an alternative `__init__` name for the member. This is not implemented in dataclasses, but is [implemented in pydantic](https://pydantic-docs.helpmanual.io/usage/schema/#field-customization). 

Add an explicit & overridable per-attribute `alias` in attrs, mirroring the behavior described in the PEP and `dataclass_transform` specification. This allows users to directly address the single largest incompatibility noted in the specification,  `attrs` [private attributes handling](https://www.attrs.org/en/stable/init.html?highlight=private#private-attributes), which strips leading underscores from member names, discussed in #795.

This feature unblocks implementation of a shim layer to support `attrs`/`dataclass` interop, discussed in:
* https://github.com/python-attrs/attrs/issues/565
* https://github.com/python-attrs/attrs/issues/686

closes: https://github.com/python-attrs/attrs/issues/945
closes: https://github.com/python-attrs/attrs/issues/572

Goals:
- [x] This is silent no-op for "normal" users of `attrs` not using `alias`. Libraries building on `attrs` *may* use `alias` to detect both existing private-name init aliases and user-supplied aliases. 
- [x] Add an `alias : Optional[str]` parameter to `attr.ib` et al and property to `Attribute`.
- [x] Lift current private-attribute-naming logic (`name.lstrip("_")`) into `_ClassBuilder` so that `alias` is always populated with the field's `__init__` parameter name, regardless of whether an explicit alias is provided, during `__init__` rendering and `attr.fields` introspection.
- [x] Leave `Attribute.alias` as `None` if not explicitly specified, so that `field_transformer` implementations can (a) create attributes without needing to be aware of `alias` and optionally provide `alias` overrides before `attrs`.
- [x] Update `attr.evolve` to use `Attribute.alias` rather than re-implemented name-mangling.
- [x] Update the attribute and private-attribute-handling documentation to note the new `alias` parameter. 
- [x] Expand dataclass_transform documentation to describe how `alias` can be used to inform type checkers of private-name renaming behavior.
- [x] Politely file a bug against `attrs` noting the inconsistency between documentation and behavior. Dunder-methods are still affected by https://github.com/python-attrs/attrs/issues/619, but `evolve` failures in pathological cases are resolved.
- [x] File a tracking issue against against `cattrs` to access `Attribute.alias` during codegen iff available in `attrs`: https://github.com/python-attrs/cattrs/issues/322

Non-goals:
- Modification of the existing private-attribute-naming logic, though could be tackled pre-release in a follow-up PR.

# Pull Request Check List

- [x] Added **tests** for changed code.
  Our CI fails if coverage is not 100%.
- [ ] New features have been added to our [Hypothesis testing strategy](https://github.com/python-attrs/attrs/blob/main/tests/strategies.py).
- [x] Changes or additions to public APIs are reflected in our type stubs (files ending in ``.pyi``).
    - [x] ...and used in the stub test file `tests/typing_example.py`.
    - [x] If they've been added to `attr/__init__.pyi`, they've *also* been re-imported in `attrs/__init__.pyi`.
- [x] Updated **documentation** for changed code.
    - [x] New functions/classes have to be added to `docs/api.rst` by hand.
    - [x] Changes to the signature of `@attr.s()` have to be added by hand too.
    - [x] Changed/added classes/methods/functions have appropriate `versionadded`, `versionchanged`, or `deprecated` [directives](http://www.sphinx-doc.org/en/stable/markup/para.html#directive-versionadded).
          Find the appropriate next version in our [``__init__.py``](https://github.com/python-attrs/attrs/blob/main/src/attr/__init__.py) file.
- [x] Documentation in `.rst` files is written using [semantic newlines](https://rhodesmill.org/brandon/2012/one-sentence-per-line/).
- [x] Changes (and possible deprecations) have news fragments in [`changelog.d`](https://github.com/python-attrs/attrs/blob/main/changelog.d).
- [x] Consider granting [push permissions to the PR branch](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork), so maintainers can fix minor issues themselves without pestering you.

### CODE REVIEW

File: docs/extending.rst [hunk 1]
```diff
@@ -260,6 +260,27 @@ A more realistic example would be to automatically convert data that you, e.g.,
...
+   ...         for field in fields
+   ...     ]
+   ...
+   >>> @frozen(field_transformer=dataclass_names)
+   ... class Data:
+   ...     public: int
+   ...     _private: float
+   ...
+   >>> Data(public=42, _private="spam")
+   Data(public=42, _privatge='spam')
```
Comment (asford) [inline docs/extending.rst:282]:
```suggestion
   Data(public=42, _private='spam')
```

File: src/attr/_funcs.py [hunk 2]
```diff
@@ -361,7 +361,7 @@ def evolve(inst, **changes):
         if not a.init:
             continue
         attr_name = a.name  # To deal with private attributes.
-        init_name = attr_name if attr_name[0] != "_" else attr_name[1:]
+        init_name = a.alias
```
Comment (hynek) [inline src/attr/_funcs.py:364]: this is actually a nice effect

File: src/attr/_make.py [hunk 3]
```diff
@@ -2247,7 +2261,10 @@ def fmt_setter_with_converter(
         has_on_setattr = a.on_setattr is not None or (
             a.on_setattr is not setters.NO_OP and has_cls_on_setattr
         )
-        arg_name = a.name.lstrip("_")
+        assert (
+            a.alias
+        ), "requires known field alias, should have been set via _ClassBuilder"
```
Comment (hynek) [inline src/attr/_make.py:2266]: did you leave this in on purpose?

File: docs/init.rst [hunk 4]
```diff
@@ -42,6 +42,7 @@ Embrace functions and classmethods as a filter between reality and what's best f
 If you look for object serialization, there's a bunch of projects listed on our ``attrs`` extensions `Wiki page`_.
 Some of them even support nested schemas.
 
+.. _private_attributes:
```
Comment (hynek) [inline docs/init.rst:45]: If I see it correctly, there's no narrative documentation for the new feature?


Would you mind adding it here and into examples.rst somewhere around the handling of private attributes?

**Review by hynek — CHANGES_REQUESTED** [review]
This looks all very good and thorough, just a bunch of questions and nits.

it also needs a newsfragment. make it sounds excited. :)

sorry for taking so long…whenever Attribute changes the diffs make me wann hide under the bed.

**Comment by hynek** [comment]
Anything you need from my end? 😇

**Review by asford — COMMENTED** [review]
My turn to apologize for the slow turnaround.

I've applied the review comments, and believe this is ready for a final pass.

File: docs/extending.rst [hunk 5]
```diff
@@ -260,6 +260,27 @@ A more realistic example would be to automatically convert data that you, e.g.,
    >>> Data(**from_json)  # ****
    Data(a=3, b='spam', c=datetime.datetime(2020, 5, 4, 13, 37))
 
+Or, perhaps you would prefer alternative default field names.
+Noting, that ``field_transformer`` receives `attrs.Attribute` instances before private-attribute handling.
```
[the remark this answers is not in the document]
Comment (asford) [reply docs/extending.rst]: Tidied and reworded.

**Comment by hynek** [comment]
w00t! Thanks for coming back! I'll try to be faster this time! :)

**Comment by asford** [comment]
Really appreciate your work on attrs.
Thanks for being stoked on the contribution!

File: src/attr/_make.py [hunk 6]
```diff
@@ -242,6 +246,7 @@ def attrib(
     .. versionchanged:: 21.1.0
        *eq*, *order*, and *cmp* also accept a custom callable
     .. versionchanged:: 21.1.0 *cmp* undeprecated
+    .. versionchanged:: 21.5.0 *alias*
```
[the remark this answers is not in the document]
Comment (hynek) [reply src/attr/_make.py:237]: you have not :D

**Review by hynek — APPROVED** [review]
Nailed it a tiny nit aside!

Thank you so much for coming back, I know how difficult it is to pick up something like this after a long time.

**Comment by Tinche** [comment]
Cool. Cattrs will be no problem, I can just bump the dependency to the latest version of attrs and implement it.

Mypy will be a bigger issue I think.

**Comment by hynek** [comment]
given alias is a dataclass transform feature, shouldn't it work OOB?

**Comment by Tinche** [comment]
Unsure, Mypy has a custom attrs plugin so I don't know if the dataclass_transform code path is in charge of this.

**Comment by hynek** [comment]
aw looks like you're right. pyright works: #1063

mypy doesn't:

Adding 

```
AliasExample(without_alias=42, _with_alias=23)
```

gives me: `tests/typing_example.py:134: error: Unexpected keyword argument "_with_alias" for "AliasExample"; did you mean "with_alias" or "without_alias"?`

@Tinche 🐶🥺?

**Comment by hynek** [comment]
yeah, it's super overdue too.

i'll try to get #1003 in ASAP (it's just an alias, so it should be just busywork) and push it out RSN.

**Pull request merged**